### ISSUE TITLE
phase property in useTransition does not work as expected

### ISSUE THREAD

**Issue by somewonderfulguy** [pmndrs/react-spring#1654]
## 🐛 Bug Report

In v8 with `useTransition` hook you could use it this way:
`{transitions.map(({item, props: {opacity}, key, state}) => { ...`
And later in the body of `map` fn use let's say such condition: `state === 'leave'`. And it works great in v8.

In v9 with `useTransition` hook there's a bit different way use:
`{transitions(({opacity}, item, {phase}) => { ...`
The `leave` in phase does not work correctly. It is no way to use `phase === 'leave'`.
In my application I develop I started migration to v9. And `leave` happens very-very rarely.

## To Reproduce

Please, see two codesanboxes I provide below.

## Expected behavior

When transitions goes away it should has `phase` as `leave`

## Link to repro

These two examples are the same, just few differences depending on `react-spring` version:

v8 - https://codesandbox.io/s/j04bi

please, see here I added `console.log` when you hover and un-hover in very calm manner, you will see only `enter` and `mount` logs. If you hover/un-hover like a frenzy, then you will see `leave` logs. Anyways, works not as expected.
v9 - https://codesandbox.io/s/leave-v9-e7072

**Comment by mjcampagna** [comment]
I wonder whether this might be related to the issue I'm having. 

I've built a few different iterations of modals, using a context and provider. Modals are in an array, allowing multiple modals to be opened simultaneously.

In the provider, I `useTransition` the modals array to mount/unmount modals using an opacity fade. This works great.

Nested within the transitioned components, the modals cards then also transition in separately, which allows different modal types to have different entry animations, for example one might slide in from the side.

This nested transition uses React state, initially set as false; I useEffect flips this to true on mount to trigger the entry animation. Again, this works great.

In my `closeModal` function, I flip this state back to false to trigger the "leave" animation, and also remove the modal from the provider array, triggering the "leave" transition for unmount. Here's the problem:

The unmount works find, and things fade out. But the nested transition fires only inconsistently. SOMETIMES, the modal will animate out, and sometimes it will not, and will only fade out with its parent.

In a previous iteration of the modals, this worked perfectly. I cannot remember the version, and I no longer have access to that code to check, but I'm now wondering if it might have been version 8.

Of course, I am now building with version 9.

I've tried nesting useSpring instead of useTransition to manage the nested animation, but it's roughly the same outcome.

Also, if I disable the wrapping transition (by not removing the modal), and simply try to "leave" animate the nested modal, it waits -- as if the wrapping transition were firing -- and then very quickly fires the nested "leave" animation after the pause, as if it's trying to catch up.

In short, it seems like the wrapping `useTransition` is eating the nested transition/spring, effectively preventing it from firing at the correct time.

Any help here would be much appreciated.
